From 9b3e3559af1ac26ce4aa096cf79911d4c28fc843 Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Tue, 9 Jan 2018 17:16:35 -0800
Subject: [PATCH 1/9] Fix unexpected behavior in SDP

Bug: 68776054
Bug: 68817966
Test: Bluetooth SDP still works
Change-Id: I4eef22679a313b88d7e8ec463b29dbb592c6b5b9
(cherry picked from commit 5d6b1b1316afecebd939f77e3d01ab0a400e68a9)
CVE-2017-13255 / CVE-2017-13256
---
 stack/sdp/sdp_server.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/stack/sdp/sdp_server.c b/stack/sdp/sdp_server.c
index 1d2f5dc..92b9d16 100644
--- a/stack/sdp/sdp_server.c
+++ b/stack/sdp/sdp_server.c
@@ -23,6 +23,8 @@
  *
  ******************************************************************************/
 
+#include <cutils/log.h>
+
 #include <stdlib.h>
 #include <string.h>
 #include <stdio.h>
@@ -348,6 +350,12 @@ static void process_service_attr_req (tCONN_CB *p_ccb, UINT16 trans_num,
         return;
     }
 
+    if (max_list_len < 4) {
+      sdpu_build_n_send_error(p_ccb, trans_num, SDP_ILLEGAL_PARAMETER, NULL);
+      android_errorWriteLog(0x534e4554, "68776054");
+      return;
+    }
+
     /* Check if this is a continuation request */
     if (*p_req)
     {
@@ -591,6 +599,12 @@ static void process_service_search_attr_req (tCONN_CB *p_ccb, UINT16 trans_num,
 
     memcpy(&attr_seq_sav, &attr_seq, sizeof(tSDP_ATTR_SEQ)) ;
 
+    if (max_list_len < 4) {
+      sdpu_build_n_send_error(p_ccb, trans_num, SDP_ILLEGAL_PARAMETER, NULL);
+      android_errorWriteLog(0x534e4554, "68817966");
+      return;
+    }
+
     /* Check if this is a continuation request */
     if (*p_req)
     {
-- 
2.7.4

